tornado_sqlalchemy/__init__.py

```diff
@@ -148,18 +126,180 @@ def _make_session(self) -> Session:
 set_max_workers = _async_exec.set_max_workers
 
 
-def make_session_factory(*args, **kwargs):
-    return SessionFactory(*args, **kwargs)
+class SessionEx(Session):
+    """The SessionEx extends the default session system with bind selection.
+    """
+
+    def __init__(self, db, autocommit=False, autoflush=True, **options):
+        self.app = db.get_app()
+        bind = options.pop('bind', None) or db.engine
+        binds = options.pop('binds', db.get_binds())
+
+        Session.__init__(
+            self, autocommit=autocommit, autoflush=autoflush,
+            bind=bind, binds=binds, **options
+        )
+
+    def get_bind(self, mapper=None, clause=None):
+        """Return the engine or connection for a given model or
+        table, using the ``__bind_key__`` if it is set.
+        """
+        # mapper is None if someone tries to just get a connection
+        if mapper is not None:
+            try:
+                # SA >= 1.3
+                persist_selectable = mapper.persist_selectable
+            except AttributeError:
+                # SA < 1.3
+                persist_selectable = mapper.mapped_table
+
+            info = getattr(persist_selectable, 'info', {})
+            bind_key = info.get('bind_key')
+            if bind_key is not None:
+                return self.app.db.get_engine(bind=bind_key)
+        return Session.get_bind(self, mapper, clause)
+
+
+class BindMeta(DeclarativeMeta):
+    def __init__(cls, name, bases, d):
+        bind_key = (
+            d.pop('__bind_key__', None)
+            or getattr(cls, '__bind_key__', None)
+        )
+
+        super(BindMeta, cls).__init__(name, bases, d)
+
+        if bind_key is not None and getattr(cls, '__table__', None) is not None:
+            cls.__table__.info['bind_key'] = bind_key
+
+
+class SQLAlchemy(object):
+
+    def __init__(self, app=None, sesion_options=None, engine_options=None):
+
+        self.Model = self.make_declarative_base()
+
+        self._engine_options = engine_options or {}
+        self._engine_lock = Lock()
+        self._engines = {}
+
+        self._sesion_options = sesion_options or {}
+
+        if app is not None:
+            self.init_app(app)
+        else:
+            self.app = None
 
+    def init_app(self, app):
+        self.app = app
+        self.app.db = self
 
-class _declarative_base:
-    def __init__(self):
-        self._instance = None
+        bind = app.settings.get('sqlalchemy_database_uri')
+        binds = app.settings.get('sqlalchemy_binds')
 
-    def __call__(self):
-        if not self._instance:
-            self._instance = sa_declarative_base()
-        return self._instance
+        if not bind and not binds:
+            raise MissingDatabaseSettingError()
 
+        engine_options = app.settings.get('sqlalchemy_engine_options') or {}
```

Comment:
This could be a one-liner (`engine_options = (app.settings.get('sqlalchemy_engine_options') or {}).copy()`).